x/iro/types/msgs.go

```diff
@@ -0,0 +1,113 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+var (
+	_ sdk.Msg = &MsgCreatePlan{}
+	_ sdk.Msg = &MsgBuy{}
+	_ sdk.Msg = &MsgSell{}
+	_ sdk.Msg = &MsgClaim{}
+	_ sdk.Msg = &MsgUpdateParams{}
+)
+
+func (m *MsgCreatePlan) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(m.Owner)
+	if err != nil {
+		return sdkerrors.ErrInvalidAddress.Wrapf("invalid owner address: %s", err)
+	}
+	plan := NewPlan(1, m.RollappId, sdk.NewCoin("dummy", m.AllocatedAmount), m.BondingCurve, m.StartTime, m.PreLaunchTime, m.IncentivePlanParams)
+	return plan.ValidateBasic()
+}
+
+func (m *MsgCreatePlan) GetSigners() []sdk.AccAddress {
+	addr := sdk.MustAccAddressFromBech32(m.Owner)
+	return []sdk.AccAddress{addr}
+}
+
+func (m *MsgBuy) ValidateBasic() error {
+	// buyer bech32
+	_, err := sdk.AccAddressFromBech32(m.Buyer)
+	if err != nil {
+		return sdkerrors.ErrInvalidAddress.Wrapf("invalid buyer address: %s", err)
+	}
+
+	// coin exist and valid
+	if m.Amount.IsNil() || m.Amount.IsZero() || m.Amount.IsNegative() {
+		return sdkerrors.ErrInvalidRequest.Wrapf("amount %v must be positive", m.Amount)
+	}
+
+	if m.ExpectedOutAmount.IsNil() || m.ExpectedOutAmount.IsZero() || m.ExpectedOutAmount.IsNegative() {
```

Comment:
same here. just `ExpectedOutAmount.IsPositive`

also for the rest of the `ValidateBasic`